### PR TITLE
[SYCL] Use a global flush buffer in stream

### PR DESCRIPTION
**Pull request by shahabl**
Change stream implementation to use a global buffer for all the
work-items to flush their outputs into. Previously a set of local flush
buffers were used for this purpose.

This change is made to avoid the issue of some devices running out of
local memory when available space is less than statement size times the
number of work-items.

Signed-off-by: Shahab Layeghi <shahab.layeghi@intel.com>

### CODE REVIEW

**Comment by s-kanaev** [comment]
Is there any impact on performance within this implementation?

**Comment by againull** [comment]
> Is there any impact on performance within this implementation?

In previous implementation where we used flush buffer in local memory users saw 10x slowdown in some cases because local memory was exhausted. So mainly this change has positive performance effect.

File: sycl/source/detail/accessor_impl.cpp [hunk 1]
```diff
@@ -22,6 +23,15 @@ AccessorImplHost::~AccessorImplHost() {
   }
 }
 
+void AccessorImplHost::resize(size_t GlobalSize) {
+  if (GlobalSize != 1) {
```
Comment (s-kanaev) [inline sycl/source/detail/accessor_impl.cpp:27]: Why 1 is special value here?

File: sycl/test/abi/symbol_size.cpp [hunk 2]
```diff
@@ -59,7 +59,7 @@ int main() {
   check_size<program, 16>();
   check_size<range<1>, 8>();
   check_size<sampler, 16>();
-  check_size<stream, 208>();
+  check_size<stream, 144>();
```
Comment (s-kanaev) [inline sycl/test/abi/symbol_size.cpp:62]: I believe you'll have to update version of library now.

File: sycl/include/CL/sycl/detail/accessor_impl.hpp [hunk 3]
```diff
@@ -79,6 +79,8 @@ class __SYCL_EXPORT AccessorImplHost {
         MElemSize(Other.MElemSize), MOffsetInBytes(Other.MOffsetInBytes),
         MIsSubBuffer(Other.MIsSubBuffer) {}
 
+  void resize(size_t GlobalSize);
```
[the remark this answers is not in the document]
Comment (shahabl) [reply sycl/include/CL/sycl/detail/accessor_impl.hpp:82]: I'll add some comment to the code here to clarify.

File: sycl/include/CL/sycl/stream.hpp [hunk 4]
```diff
@@ -126,19 +126,9 @@ class __SYCL_EXPORT stream {
...
-  // Each work item in the work group writes to its own flush buffer in the
-  // pool. This accessor is used to atomically get offset of the flush buffer in
-  // the pool for each work item in the work group. This approach is used
-  // because currently it is not possible to get work item id in the work group
-  // without id object, which is passed to the kernel.
-  mutable detail::stream_impl::LocalOffsetAccessorT WIOffsetAcc;
-
-  mutable detail::stream_impl::GlobalOffsetAccessorT FlushSize;
+  // Accessor to the flush buffer. Each work item writes its
+  // output to a designated section of the flush buffer.
```
[the remark this answers is not in the document]
Comment (shahabl) [reply sycl/include/CL/sycl/stream.hpp:130]: The write methods and other methods that update the flush buffer check the size and won't write beyond the dedicated range. So if someone tries to write to the stream more than the space requested for the statement at stream creation, the extra characters will be ignored.

File: sycl/source/detail/accessor_impl.cpp (hunk 1 above)
Comment (shahabl) [reply sycl/source/detail/accessor_impl.cpp:27]: No special significance anymore, but now it prevents some unnecessary operations.

File: sycl/test/abi/symbol_size.cpp (hunk 2 above)
Comment (shahabl) [reply sycl/test/abi/symbol_size.cpp:62]: Is there something besides the change already done in sycl/CMakeLists.txt?

Comment (s-kanaev) [reply sycl/test/abi/symbol_size.cpp:62]: Please, refer to this document: https://github.com/intel/llvm/blob/sycl/sycl/doc/ABIPolicyGuide.md

File: sycl/source/detail/accessor_impl.cpp (hunk 1 above)
Comment (s-kanaev) [reply sycl/source/detail/accessor_impl.cpp:27]: OK, then, when do you call `resize` method with `GlobalSize` equal to one?

**Comment by againull** [comment]
@s-kanaev, @v-klochkov  Merging is blocked, I believe because your approve is required on behalf of runtime reviewers. Please approved or provide your feedback.

**Pull request merged**